src/main/java/frc/robot/swervev2/SwervePosEstimator.java

```diff
@@ -7,40 +7,54 @@
 import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
 import edu.wpi.first.math.kinematics.SwerveModulePosition;
 import edu.wpi.first.math.numbers.N3;
-import edu.wpi.first.math.util.Units;
-import edu.wpi.first.networktables.DoubleArraySubscriber;
-import edu.wpi.first.networktables.NetworkTable;
-import edu.wpi.first.networktables.NetworkTableInstance;
-import edu.wpi.first.networktables.TimestampedDoubleArray;
-import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.networktables.*;
 import edu.wpi.first.wpilibj.Timer;
 import edu.wpi.first.wpilibj.smartdashboard.Field2d;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.Robot;
 import frc.robot.constants.Constants;
 import frc.robot.swervev2.components.GenericEncodedSwerve;
+import frc.robot.utils.Apriltag;
+import frc.robot.utils.PrecisionTime;
+import frc.robot.utils.RobotMode;
+import frc.robot.utils.math.ArrayUtils;
+import frc.robot.utils.math.PoseUtils;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Class to estimate the current position of the robot,
  * given the serve wheel encoders (through swerve model states)
  * and the initial position of the robot
  */
-public class SwervePosEstimator{
+public class SwervePosEstimator {
+    public record VisionMeasurement(TimestampedDoubleArray measurement, Apriltag tag, double timeOfMeasurement){}
     private final Field2d field = new Field2d();
     private final GenericEncodedSwerve frontLeftMotor;
     private final GenericEncodedSwerve frontRightMotor;
     private final GenericEncodedSwerve backLeftMotor;
     private final GenericEncodedSwerve backRightMotor;
     private final SwerveDrivePoseEstimator poseEstimator;
-    private final DoubleArraySubscriber subscriber;
+    private final DoubleArraySubscriber visionMeasurementSubscriber;
+    private final IntegerArraySubscriber apriltagIdSubscriber;
 
     /* standard deviation of robot states, the lower the numbers arm, the more we trust odometry */
-    private static final Vector<N3> stateStdDevs = VecBuilder.fill(0.01, 0.01, 0.001);
+    private static final Vector<N3> stateStdDevs = VecBuilder.fill(0.15, 0.15, 0.001);
 
     /* standard deviation of vision readings, the lower the numbers arm, the more we trust vision */
-    private static final Vector<N3> visionMeasurementStdDevs = VecBuilder.fill(0.7, 0.7, 0.3);
+    private static final Vector<N3> visionMeasurementStdDevs = VecBuilder.fill(0.4, 0.4, 0.5);
+    private static final Transform2d cameraOneTransform = new Transform2d(Constants.CAMERA_OFFSET_FROM_CENTER_X, Constants.CAMERA_OFFSET_FROM_CENTER_Y, new Rotation2d());
+    private static final Transform2d cameraTwoTransform = new Transform2d(Constants.CAMERA_OFFSET_FROM_CENTER_X, Constants.CAMERA_OFFSET_FROM_CENTER_Y, new Rotation2d());
```

Comment:
it is the same because this code was going to account for two camera but we did not get there (also we will problably use ros for the offset in the future)